backend/src/stwo/circuit_builder.rs

```diff
@@ -0,0 +1,269 @@
+                    .collect()
+            })
+            .unwrap()
+    }
+}
+
+#[derive(Clone)]
+//maybe change the name to PowdrEval, follows stwo tradition
+//to do, witness related should move to powdrcircuit
+pub(crate) struct PowdrEval<'a, T> {
+    analyzed: Arc<Analyzed<T>>,
+    /// Callback to augment the witness in the later stages.
+    witgen_callback: Option<WitgenCallback<T>>,
+    /// The value of the witness columns, if set
+    witness: Option<&'a [(String, Vec<T>)]>,
+    pub log_n_rows: u32,
+    // pub lookup_elements: LookupElements<2>,
+    // pub claimed_sum: ClaimedPrefixSum,
+    // pub total_sum: SecureField,
+    // pub base_trace_location: TreeSubspan,
+    // pub interaction_trace_location: TreeSubspan,
+    // pub constants_trace_location: TreeSubspan,
+}
+
+impl<'a, T: FieldElement> PowdrEval<'a, T> {
+    pub(crate) fn new(analyzed: Arc<Analyzed<T>>) -> Self {
+        let degree_log = analyzed.degree().ilog2();
+        Self {
+            analyzed,
+            witgen_callback: None,
+            witness: None,
+            log_n_rows: degree_log,
+            // lookup_elements: unimplemented!(),
+            // claimed_sum: unimplemented!(),
+            // total_sum: unimplemented!(),
+            // base_trace_location: unimplemented!(),
+            // interaction_trace_location: unimplemented!(),
+            // constants_trace_location: unimplemented!(),
+        }
+    }
+
+    pub(crate) fn with_witgen_callback(self, witgen_callback: WitgenCallback<T>) -> Self {
+        Self {
+            witgen_callback: Some(witgen_callback),
+            ..self
+        }
+    }
+
+    pub(crate) fn with_witness(self, witness: &'a [(String, Vec<T>)]) -> Self {
+        Self {
+            witness: Some(witness),
+            ..self
+        }
+    }
+}
+
+impl<'a, T: FieldElement> FrameworkEval for PowdrEval<'a, T> {
+    fn log_size(&self) -> u32 {
+        self.log_n_rows
+    }
+    fn max_constraint_log_degree_bound(&self) -> u32 {
+        self.log_n_rows + 1
+    }
+    fn evaluate<E: EvalAtRow>(&self, mut eval: E) -> E {
+        // Add polynomial identities
+        let identities = self
+            .analyzed
+            .identities_with_inlined_intermediate_polynomials()
+            .into_iter()
+            .filter(|id| id.kind == IdentityKind::Polynomial)
+            .collect::<Vec<_>>();
+        if !identities.is_empty() {
+            identities.iter().for_each(|id| {
+                let expr = id.expression_for_poly_id();
+                let name = id.to_string();
+                println!(
+                    "\n this is the name {:?}, this is the expr {:?} \n",
+                    name, expr
+                );
+                let expr = to_stwo_expression(expr, &mut eval);
+                eval.add_constraint(expr);
+            });
+        }
+        eval
+    }
+}
+
+fn to_stwo_expression<T: FieldElement, E: EvalAtRow>(
+    expr: &AlgebraicExpression<T>,
+    eval: &mut E,
```

Comment:
I think this cannot take a mutable reference: if we call this function twice on witness x, we want to receive the same stwo witness. Now it would create a new witness each time. My intuition is that we should first allocate all witnesses into a vector by calling next_interaction as many times as needed. Then in this function when we see a reference, we get the associated stwo reference. The plonky3 code can be helpful here: the vector of plonky3 witnesses is returned by the builder. In stwo, we have to build it ourselves using E.